app/master/cluster_master.py

```diff
@@ -62,12 +65,24 @@ def api_representation(self):
             'slaves': slaves_representation,
         }
 
-    def builds(self):
+    def builds(self, offset: int=None, limit: int=None) -> List['Build']:
         """
-        Returns a list of all builds
-        :rtype: list[Build]
+        Returns a list of all builds. Both offset and limit are assumed to be nonnegative at this point
+        since the corrections or negative offsets and limits are done when the query parameters are first initialized.
+        :param offset: The starting index of the requested build
+        :param limit: The number of builds requested
         """
-        return self._all_builds_by_id.values()
+        num_builds = len(self._all_builds_by_id)
+        offset = offset or 0
+        # Don't return more than the MAX_LIMIT even if we have more builds than that
+        limit = limit if limit is not None else min(num_builds, pagination_constants['MAX_LIMIT'])
+
+        # Reset the offset to 0 if it is out of range
+        starting_index = offset if offset < num_builds else 0
+        ending_index = min((starting_index + limit), num_builds)
+
+        requested_builds = islice(self._all_builds_by_id, starting_index, ending_index)
```

Comment:
That's a good catch -- alternatively we could create a range of build ids from the offset/limit and explicitly retrieve those from the OrderedDict. 

Like `offset=5&limit=20` would imply we want builds 6 to 25, so we could just fetch all the builds with an id from 6 to 25 from the OrderedDict (we could then just use a regular dict)